alphagov/paas-docker-cloudfoundry-tools
# [#123003931] Add concourse `enum` resource type

## What

[#123003931](https://www.pivotaltracker.com/n/projects/1275640/stories/123003931) Improve "User" (PaaS Operator) Experience for creating  new Environment

Introduces a new concourse enum resource type.

This type enables us to model the state of an enum backed in a s3 bucket.  Its key improvements are to encapsulate a pattern we found useful while implementing [#123003931](https://www.pivotaltracker.com/n/projects/1275640/stories/123003931)

## How to review

* Checkout this branch and modify the example pipeline to point at the docker tag for this branch:

~~~
diff --git a/enum-resource/example_pipeline.yml b/enum-resource/example_pipeline.yml
index 5b0b624..ad496f5 100644
--- a/enum-resource/example_pipeline.yml
+++ b/enum-resource/example_pipeline.yml
@@ -4,6 +4,7 @@ resource_types:
   type: docker-image
   source:
     repository: governmentpaas/enum-resource
+    tag: feature_123003931_concourse_enum_type

 resources:
 - name: sample-enum
~~~

* Upload the pipeline to a development concourse

~~~
fly set-pipeline -p example-enum -c example_pipeline.yml -v bucket-name=${DEPLOY_ENV}-state -v aws_region=eu-west-1
~~~

* Run the jobs `sample_enum_get_job` and `sample_enum_set_job` and see the enum flips from a default value (`no`) to a set value (`yes`)
* You can also check against the state in the s3 bucket

~~~
aws s3 cp s3://${DEPLOY_ENV}-state/sample-enum -
~~~

## Who can review

Not @richardc @benhyland
